tools/detekt/src/main/kotlin/com/datadog/tools/detekt/rules/sdk/PackageNameVisibility.kt

```diff
@@ -0,0 +1,101 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.tools.detekt.rules.sdk
+
+import com.datadog.tools.detekt.rules.AbstractTypedRule
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Severity
+import io.gitlab.arturbosch.detekt.api.config
+import io.gitlab.arturbosch.detekt.api.internal.RequiresTypeResolution
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtModifierListOwner
+import org.jetbrains.kotlin.psi.KtNamedDeclaration
+import org.jetbrains.kotlin.psi.KtPackageDirective
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.util.isAnnotated
+
+/**
+ * A rule to detekt classes in the wrong package name.
+ *
+ * Our naming policy implies that a class that is Internal to the SDK be in a
+ * `com.datadog.android.*.internal.*` package name.
+ * @active
+ */
+@RequiresTypeResolution
+class PackageNameVisibility(
+    ruleSetConfig: Config
+) : AbstractTypedRule(ruleSetConfig) {
+
+    private val withBreakingChanges: Boolean by config(defaultValue = true)
+    private val ignoredAnnotations: List<String> by config(defaultValue = emptyList())
+
+    override val issue: Issue = Issue(
+        javaClass.simpleName,
+        Severity.Warning,
+        "This rule reports when a top level declaration's visibility doesn't match the package naming policy.",
```

Comment:
semantically we then have a problem with classes generated from `@NoOpImplementation` declaration - they will be `internal` by default, but in the same package (which may be non-internal) as the original interface.